<a href="https://colab.research.google.com/github/Indrejue/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/Antony_Farag__LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [19]:
# YOUR WORK HERE
#load libraries needed for working
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import chi2_contingency


In [4]:
#load in student score files
math = pd.read_csv('student-mat.csv',sep=';')
print(math.shape)
math.head()

(395, 33)


school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [ ]:
# YOUR WORK HERE
#variables that could have a high corelation study time and failures

hypothisis= no association between absences and failure rates

alternative hypothisis = likely association between absences and failure rates

In [6]:
#take a quick look at failure rates
fail = pd.crosstab (index=math['failures'],columns='count')
fail

col_0     count
failures       
0           312
1            50
2            17
3            16

In [10]:
# take a quick look at absenties
absen= pd.crosstab (index=math['studytime'],columns='count')
absen

col_0      count
studytime       
1            105
2            198
3             65
4             27

In [26]:
# look at total failures compared to study times
pd.crosstab(math['failures'],math['studytime'],margins=True)

studytime    1    2   3   4  All
failures                        
0           74  158  54  26  312
1           16   26   7   1   50
2            6    7   4   0   17
3            9    7   0   0   16
All        105  198  65  27  395

In [27]:
# look at total failures compared to sudy times as a percent
pd.crosstab(math['failures'],math['studytime'],normalize='index')*100

studytime          1          2          3         4
failures                                            
0          23.717949  50.641026  17.307692  8.333333
1          32.000000  52.000000  14.000000  2.000000
2          35.294118  41.176471  23.529412  0.000000
3          56.250000  43.750000   0.000000  0.000000

In [32]:
# do a chi squared test to see if there is association or not between study time and failure rate
chi1, pval1, dof1, expected1 = chi2_contingency(pd.crosstab(math['famsize'],math['studytime']))
print (chi1)
print (pval1)
print (dof1)
print (expected1)

16.21199080868576
0.06258448399974005
9
[[ 82.93670886 156.39493671  51.34177215  21.32658228]
 [ 13.29113924  25.06329114   8.2278481    3.41772152]
 [  4.51898734   8.52151899   2.79746835   1.16202532]
 [  4.25316456   8.02025316   2.63291139   1.09367089]]


based on a p-value of .0626 we fail to reject the hypothisis that there is no relation between study time and failure rates.

In [ ]:
#variables that should not have a high corelation travel time and free time

In [33]:
chi1, pval1, dof1, expected1 = chi2_contingency(pd.crosstab(math['traveltime'],math['freetime']))
print (chi1)
print (pval1)
print (dof1)
print (expected1)

11.46774269441963
0.4893142424435174
12
[[ 12.36202532  41.64050633 102.14936709  74.82278481  26.02531646]
 [  5.14683544  17.33670886  42.52911392  31.15189873  10.83544304]
 [  1.10632911   3.72658228   9.14177215   6.69620253   2.32911392]
 [  0.38481013   1.29620253   3.17974684   2.32911392   0.81012658]]


### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [ ]:
# YOUR WORK HERE

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [ ]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)